docs/source/hooks.rst

```diff
@@ -0,0 +1,160 @@
+Hooks make it possible to track state changes to each job in a **signac** project as the :ref:`FlowProject<flow-project>` operates on it.
+
+A hook is a function that is called at a specific point relative to the execution of a **signac-flow** :ref:`operation<operations>`.
+A hook is triggered when an operation starts, exits, succeeds, or fails.
+
+A basic use case is to log the success/failure of an operation by creating a hook that sets a job document value ``job.doc.operation_success`` to ``True`` or ``False``.
+Additionally, a user may record the `git commit ID <https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History>`_ upon the start of an operation, allowing them to track which version of code ran the operation.
+
+.. _hook_triggers:
+
+Hook Triggers
+=============
+
+The following triggers are provided:
+
+1. :py:meth:`~flow.FlowProject.add_hook.on_start` will execute when the operation begins execution.
+2. :py:meth:`~flow.FlowProject.add_hook.on_exit` will execute when the operation exits, with or without error.
+3. :py:meth:`~flow.FlowProject.add_hook.on_success` will execute when the operation exits without error.
+4. :py:meth:`~flow.FlowProject.add_hook.on_error` will execute when the operation exits with error.
+
+Hooks can be installed at the :ref:`operation level <operation hooks>` or at the :ref:`flow-project level<project-level hooks>`.
+Project-level hooks are called for every operation in the flow project.
+
+.. note::
+
+    Hooks are run in the environment of the python process from which you call FlowProject.main().
+    For this reason, hooks will not have access to modules in a container specified in the :term:`executable directive<executable>`.
+
+.. _operation hooks:
+
+Operation Hooks
+===============
+
+Hooks may be added to individual operations using decorators.
+The :py:class:`~flow.FlowProject.add_hook` decorator tells :py:class:`~signac` to run a hook (or set of hooks) when an operation reaches the specified trigger.
+
+The :py:class:`~flow.FlowProject.add_hook` decorator accepts objects as a function of the job operation (:py:class:`~flow.project.JobOperation`).
+
+
+The decorators :py:meth:`~flow.FlowProject.add_hook.on_start` and  :py:meth:`~flow.FlowProject.add_hook.on_exit` accept functions with two parameters: the operation name and the :py:class:`Job` object.
+The decorator :py:meth:`~flow.FlowProject.add_hook.on_error`, accepts functions with three parameters: the operation name, the output error, and the :py:class:`Job` object.
+
+Unlike :py:meth:`~flow.FlowProject.add_hook.on_start`, :py:meth:`~flow.FlowProject.add_hook.on_exit`,  and :py:meth:`~flow.FlowProject.add_hook.on_success`, which accept functions that take ``operation_name`` and ``job`` as arguments, :py:meth:`~flow.FlowProject.add_hook.on_error` accepts functions that take ``operation_name``, ``error``, and ``job`` as arguments.
```

Comment:
The bottom version was in a "note", but these are basically the same info.

Because the number of arguments matters for project-level-hooks too, I want to condense these and move it to the section that talks about kinds of triggers available.

Reply:
See also the draft text I added to the trigger section